packages/alfa-rules/src/sia-r15/rule.ts

```diff
@@ -107,3 +106,71 @@ export namespace Outcomes {
     )
   );
 }
+
+type Source = "srcdoc" | "src" | "invalid" | "nothing";
+
+class EmbeddedResource implements Equatable, Hashable {
+  public static of(source: Source, value?: string): EmbeddedResource {
+    return new EmbeddedResource(source, Option.from(value));
+  }
+
+  private readonly _source: Source;
+  private readonly _value: Option<string>;
+
+  private constructor(source: Source, value: Option<string>) {
+    this._source = source;
+    this._value = value;
+  }
+
+  public get value(): Option<string> {
+    return this._value;
+  }
+
+  public equals(value: unknown): value is this {
+    return (
+      value instanceof EmbeddedResource &&
+      this._source === value._source &&
+      this._value.equals(value._value)
+    );
+  }
+
+  public hash(hash: Hash) {
+    Hash.writeString(hash, this._source);
+    this._value.hash(hash);
+  }
+
+  public toString() {
+    return `{source: ${this._source}, value: ${this._value}}`;
+  }
+}
```

Comment:
This class looks like a rather complicated wrapper around a `string` and so whatever it solves could just as well be solved by a `Map` 🤔

Reply:
🤔 this is a couple of `string` where the second one may be missing. Structural equality is needed, so `[string, string]` is not really good. I'm not sure a `Map` is the best way to model that (as in, in every case there will always be exactly one key with a value).

Reply:
See https://github.com/Siteimprove/alfa/pull/406#discussion_r497617734.